pkg/logcli/query/tail.go

```diff
@@ -44,9 +42,32 @@ func (q *Query) TailQuery(delayFor time.Duration, c client.Client, out output.Lo
 		log.Println("Print only labels key:", color.RedString(strings.Join(q.ShowLabelsKey, ",")))
 	}
 
+	tailResponse := new(loghttp.TailResponse)
+	lastReceivedTimestamp := q.Start
+
 	for {
 		err := unmarshal.ReadTailResponseJSON(tailResponse, conn)
 		if err != nil {
+			// Check if the websocket connection closed unexpectedly. If so, retry.
+			// The connection might close unexpectedly if the querier handling the tail request
+			// in Loki stops running. The following error would be printed:
+			// "websocket: close 1006 (abnormal closure): unexpected EOF"
+			if websocket.IsCloseError(err, websocket.CloseAbnormalClosure) {
+				log.Printf("Remote websocket connection closed unexpectedly (%+v). Connecting again.", err)
+
+				// Close previous connection
+				if err = conn.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(websocket.CloseNormalClosure, "")); err != nil {
+					log.Fatalf("Error closing websocket: %+v", err)
+				}
+
+				conn, err = c.LiveTailQueryConn(q.QueryString, delayFor, q.Limit, lastReceivedTimestamp, q.Quiet)
+				if err != nil {
+					log.Fatalf("Error recreating tailing connection after unexpected close: %+v", err)
```

Comment:
Shouldn't we break here?

Reply:
Done at [24260d1](https://github.com/grafana/loki/pull/8374/commits/24260d11a4d70753326d9c833f1102efb8cd5bb8)